### ISSUE TITLE
Streaming response continuously?

### ISSUE THREAD

**Issue by thojanssens** [keathley/finch#58]
Imagine connecting to a forex/crypto trader that streams the response continuously with real-time price changes.

I can achieve that with Mint easily through their API (see sample code below); we can easily stream and read from the response forever. But it seems that it is not possible with Flinch?

I stumbled upon the same demand for Tesla:
https://github.com/teamon/tesla/issues/271
but open since 2018

Problem with mint is that I don't have a connection pooling (there's mint_pool in development but stopped for now) and using something higher-level such as flinch would be easier to work with.

```
defmodule Forex do
  defp receive_next_and_stream(conn) do
    receive do
      message ->
        result = Mint.HTTP.stream(conn, message)
        {:ok, conn, responses} = result
        receive_next_and_stream(conn)
    end
  end

  def stream_prices() do
    {:ok, conn} =
      Mint.HTTP.connect(:https, "stream-fxpractice.oanda.com", 443, timeout: :infinity)

    uri = "/v3/accounts/" <> account_id <> "/pricing/stream?instruments=EUR_GBP"

    {:ok, conn, request_ref} = HTTP.request(conn, "GET", uri, [{"Authorization", token}])

    receive_next_and_stream(conn)
  end
end
```

**Comment by keathley** [comment]
We've talked about adding streaming api support to both http/1.1 and http/2 pools. It just hasn't been a priority yet. We're not against the idea though. Is the goal to open a connection to a url and continually stream responses to the calling process? Or do you need something more advanced than that?

**Comment by lessless** [comment]
In my case it would be exactly the first scenario - continuously receiving an unbound stream of updates.   
From about 5-6 different endpoints on a same host.

**Comment by thojanssens** [comment]
What do you have in mind Chris when you mention something more advanced? Just curious about your ideas.

I want to add something, and I'll keep the example of brokers for Forex exchanges or crypto. Some send market price updates by continuously streaming HTTP responses on a connection, as described in this issue; but other brokers choose to send those price updates through WebSocket protocol. Would that also be the job for an HTTP client such as Finch? I think so; but if not, why and what tool do I need to use to get those price updates through a WebSocket?

**Comment by keathley** [comment]
@thojanssens If you need bi-directional communication that is guaranteed to be on the same connection, you can't do that with http/1.1. It's only possible on http/2. Along with that limitation, our current pooling/load balancing strategy doesn't give you control over which connection your request uses. For instance, if you start 4 http/2 connections and you begin streaming a response on one connection if you create a new request that request can end up on any of the 4 connections.

I don't think the use case I just described is something that we're going to attempt to solve in Finch. If you need to have that much control you're better off using Mint directly. But I think Finch can provide mechanisms for streaming responses. We'll need to provide a caveat for http/1.1 pools that if you start streaming responses with limited pool sizes you may end up exhausting your pool. But anyone who needs streaming in that way is most likely aware of the tradeoff.

**Comment by keathley** [comment]
Streaming support has been merged with #59. If y'all would like to try it out that would be great. I'm going to close this issue for now. If there are specific issues that come up please let us know.
